### PR TITLE
MCO-1304: Add boot image update e2e tests

### PR DESCRIPTION
**Pull request by djoshy**
This PR adds e2e tests for GCP(in GA) and AWS(in tech preview) boot image updates. 

This is also the first set of e2e tests added by the MCO team to origin, so I have added a new OWNERS file for the package.

### CODE REVIEW

**Review by yuqi-zhang — COMMENTED** [review]
Generally lgtm! Just some questions inline.

Also, would we eventually plan on testing bootimage updates on old installs somehow? It's not necessary for this test, but am curious what you think would make the most sense to test

File: test/extended/machine_config/boot_image_update_aws.go [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+		// Clear out boot image configuration between tests
+		err := oc.Run("apply").Args("-f", noneMachineSetFixture).Execute()
+		o.Expect(err).NotTo(o.HaveOccurred())
+	})
+
+	g.It("Should update boot images only on MachineSets that are opted in [apigroup:machineconfiguration.openshift.io]", func() {
+		PartialMachineSetTest(oc, partialMachineSetFixture)
+	})
+
+	g.It("Should update boot images on all MachineSets when configured [apigroup:machineconfiguration.openshift.io]", func() {
```
Comment (sosiouxme) [inline test/extended/machine_config/boot_image_update_aws.go:42]: FWIW this one [failed](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/29099/pull-ci-openshift-origin-master-e2e-aws-ovn-single-node-techpreview-serial/1838305875253530624)

**Review by sosiouxme — COMMENTED** [review]
/approve

**Comment by sosiouxme** [comment]
BTW I don't think I'd worry too much about `[sig-architecture] platform pods should not exit more than once with a non-zero exit code` which is flaky but succeeds enough that retries always get it across the line. i'd worry more about `[sig-arch] events should not repeat pathologically for ns/openshift-ovn-kubernetes` which does actually succeed all the time.

File: test/extended/machine_config/boot_image_update_gcp.go [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+	osconfigv1 "github.com/openshift/api/config/v1"
+
+	g "github.com/onsi/ginkgo/v2"
+	o "github.com/onsi/gomega"
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+// This test is [Serial] because it modifies the cluster/machineconfigurations.operator.openshift.io object in each test.
+var _ = g.Describe("[sig-coreos][OCPFeatureGate:ManagedBootImages][Serial]", func() {
```
[the remark this answers is not in the document]
Comment (djoshy) [reply test/extended/machine_config/boot_image_update_gcp.go:15]: No opposition from me (: will update to include it

File: test/extended/machine_config/helpers.go [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+// verifyMachineSetUpdate verifies that the the boot image values of a MachineSet are reconciled correctly
+func verifyMachineSetUpdate(oc *exutil.CLI, machineSet machinev1beta1.MachineSet, updateExpected bool) {
+
+	newProviderSpecPatch, originalProviderSpecPatch, newBootImage, originalBootImage := createFakeUpdatePatch(oc, machineSet)
+	err := oc.Run("patch").Args(mapiMachinesetQualifiedName, machineSet.Name, "-p", fmt.Sprintf(`{"spec":{"template":{"spec":{"providerSpec":{"value":%s}}}}}`, newProviderSpecPatch), "-n", mapiNamespace, "--type=merge").Execute()
+	o.Expect(err).NotTo(o.HaveOccurred())
+
+	// Ensure atleast one master node is ready
+	framework.Logf("Waiting until atleast one master node is ready...")
+	WaitForOneMasterNodeToBeReady(oc)
```
[the remark this answers is not in the document]
Comment (djoshy) [reply test/extended/machine_config/helpers.go:95]: So this is one of the preconditions before updating boot images to ensure that the MCO itself is upto date and not in the middle of an update: https://github.com/openshift/machine-config-operator/blob/master/pkg/operator/sync.go#L1825-L1826 

This is what stamps the configmap, which lets the MSBIC continue. This check is unnecessary here as the test is not during an upgrade, but I just added it for safety. Perhaps I should remove it?

File: test/extended/machine_config/boot_image_update_aws.go (hunk 1 above)
Comment (djoshy) [reply test/extended/machine_config/boot_image_update_aws.go:42]: This is strange...looking at the logs it seems like the MachineConfiguration object was out of date when the MCO's controller was trying to roll out the boot image update(resulting in the update not taking place). Perhaps some kind of SNO unstability, resulting in slow API requests? 
```
E0923 23:08:14.916835       1 leaderelection.go:340] Failed to update lock optimitically: Put "https://172.30.0.1:443/apis/coordination.k8s.io/v1/namespaces/openshift-machine-config-operator/leases/machine-config-controller": dial tcp 172.30.0.1:443: connect: connection refused, falling back to slow path
E0923 23:08:14.917493       1 leaderelection.go:347] error retrieving resource lock openshift-machine-config-operator/machine-config-controller: Get "https://172.30.0.1:443/apis/coordination.k8s.io/v1/namespaces/openshift-machine-config-operator/leases/machine-config-controller": dial tcp 172.30.0.1:443: connect: connection refused
```
It almost seems like the node was in the middle of an update which doesn't make sense :thinking:

**Comment by djoshy** [comment]
> BTW I don't think I'd worry too much about [sig-architecture] platform pods should not exit more than once with a non-zero exit code which is flaky but succeeds enough that retries always get it across the line. i'd worry more about [sig-arch] events should not repeat pathologically for ns/openshift-ovn-kubernetes which does actually succeed all the time.

ack, will take a closer look at this failure! Thanks!

File: test/extended/machine_config/helpers.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (yuqi-zhang) [reply test/extended/machine_config/helpers.go]: I think it's fine, I didn't draw the connection :)

**Comment by djoshy** [comment]
>Also, would we eventually plan on testing bootimage updates on old installs somehow? It's not necessary for this test, but am curious what you think would make the most sense to test

Missed this on my last pass, sorry. I think the most logical place would be a post-upgrade test(when there is an actual boot image diff). The tricky thing is that I don't think there is an existing workflow for that in tech preview. We could start with adding them for GCP since that has GAed already.

**Comment by sosiouxme** [comment]
The tests you added are running successfully on multiple variants.
Other stuff is going wrong in the tests but I can't see how your additions could possibly have affected them.
I'm good with overriding the failures... everything but e2e-gcp-ovn-builds would probably eventually pass, but what's the point.

/override ci/prow/e2e-gcp-ovn-builds
/override ci/prow/e2e-gcp-fips-serial
/override ci/prow/e2e-aws-ovn-ipsec-serial
/override ci/prow/e2e-aws-ovn-single-node-techpreview-serial
/override ci/prow/e2e-aws-ovn-single-node-upgrade

**Pull request merged**